### ISSUE TITLE
Delays start 1 turn too late

### ISSUE THREAD

**Issue by powerbf** [crawl/crawl#2376]
Trunk. Any platform

The function start_delay does not actually start the delay. It just pushes it into the queue, and it gets picked up and started on the next turn. This has two effects:

1. The delay is one turn longer than expected. This is compensated for in some calls by subtracting one from the delay length, but this is not done consistently, and in any case, it's weird and counter-intuitive.
2. Things can happen between an action being taken and it actually starting. For example, when casting Passwall, monsters can take a swing at you between casting the spell and getting the message "You begin to meditate on the wall."

**Comment by rawlins** [comment]
>The function start_delay does not actually start the delay. It just pushes it into the queue, and it gets picked up and started on the next turn.

Agreed that delays are complicated, but overall this is an intentional part of how delays work, in that all delay handling happens in a specific timing in the game loop sequence for the sake of stable assumptions about game state, reentry, etc. It's not quite right that they all get picked up on the next turn: delays that don't end the turn (`you.turn_is_over`, e.g. the code path for macros does not set this) are processed immediately, but if they do end the turn (e.g. by being triggered while casting a spell) they happen right after `world_reacts`, which does indeed give monsters an initial chance to interrupt. I'm not sure it's completely out to move the call to `handle_delay` earlier in the game loop, but it seems extremely tricky to get the timing issues right when doing that.

For passwall in particular after some `#crawl-dev` discussion we agree that it involves a confusing message order, and could potentially be fixed by doing the `::start()` call when the delay is pushed (FWIW the `::start()` calls are currently only messaging so calling that earlier is irrelevant to the delay length). However, it's not done this way in general because of other message ordering issues. The [original bug](https://sourceforge.net/p/crawl-ref/bugs/559/) that led to this change is no longer relevant (props to @ebering for even finding that), but on some quick testing that @ebering did there seem to be some subtleties because of the delays that don't set `you.turn_is_over`. So possibly the solution is to handle its messaging outside of `::start()` altogether, rather than changing delay behavior. Something similar applies to shaft self.

> 1. The delay is one turn longer than expected.  This is compensated for in some instances by subtracting one from the delay length

Re 1, are there other specific delays that you think are incorrect? I.e. are actually one turn longer than intended?

**Comment by powerbf** [comment]
Thanks for the reply.

> Re 1, are there other specific delays that you think are incorrect? I.e. are actually one turn longer than intended?

On taking a second look, the only real inconsistency is in item-use.cc: _puton_amulet uses ARMOUR_EQUIP_DELAY, but takeoff_armour and remove_ring (when handling remove amulet) use ARMOUR_EQUIP_DELAY - 1. I thought the -1 was there to compensate for start_delay not starting immediately, but maybe it should just be a different constant.
